horovod/common/util.py

```diff
@@ -251,3 +253,11 @@ def split_list(l, n):
     """
     d, r = divmod(len(l), n)
     return [l[i * d + min(i, r):(i + 1) * d + min(i + 1, r)] for i in range(n)]
+
+
+def check_framework_version(name, version):
```

Comment:
See comment in TensorFlow call site for more details, but basically I think we can add an additional optional parameter here:

```
def check_framework_version(name, version, exception):
```

Then we handle a version mismatch differently depending on if we're handling an exception or not.  If we're handling an exception, we want to wrap and re-raise the exception.  If we're not handling an exception, we want to issue a warning.

So:

```
...
if installed_version != version:
    if exception is not None:
        raise HorovodVersionMismatchError(name, version, installed_version) from exception
    else:
        warnings.warn(..., HorovodVersionMismatchWarning)
```

What do you think?